### PR TITLE
Return unsafe_view instead of view from matmul when folding occurs

### PR DESCRIPTION
**Pull request by jwieczorekhabana**
When tensor folding occurs during matmul operation returned tensor is a view. This can cause issues when matmul is used inside a custom function and such view is then returned as output. Then it cannot be modified inplace and causes errors.
It can be especially problematic when after such function inplace allreduce is performed.
Issue is resolved when unsafe_view is returned from matmul instead. This solution aligns matmul decomposition with eager implementation in such a way that a non view tensor is returned.

Test included in this PR reproduces the issue.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

**Comment by xinyu-intel** [comment]
@jgong5 @EikanWang Hi, can you help on review the PR?

**Comment by jwieczorekhabana** [comment]
Error thrown:
image

**Comment by jwieczorekhabana** [comment]
@zou3519 Hi I have linter patch ready. In the error messages I've also seen note about adding test owner. From what I've seen in similar test it's just '# Owner(s): ["module: unknown"]. I'm also wondering whether it's the right place for this kind of test or you might already have a more fitted suite somewhere else. Could you give any hints on that?

File: test/custom_function/test_custom_function.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    def test_autograd_function_with_matmul_folding_at_output(self):
+        """
+        When tensor folding occurs during matmul operation returned tensor is a view.
+        This can cause issues when matmul is used inside a custom function
+        and such view is then returned as output. Then it cannot be modified inplace
+        and causes errors.
+        It can be especially problematic when after such function inplace allreduce
+        is performed. This test recreates this behaviour.
+        Issue is resolved when unsafe_view is returned from matmul instead.
+        """
```
Comment (zou3519) [inline test/custom_function/test_custom_function.py:17]: Move this to test/dynamo/test_misc.py

**Review by zou3519 — COMMENTED** [review]
The code change makes sense to me. Let's move the test to an existing file rather than put it in its own testcase.

**Review by zou3519 — APPROVED** [review]

**Review by zou3519 — CHANGES_REQUESTED** [review]
lint failing

**Comment by jwieczorekhabana** [comment]
> lint failing

done

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**